lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -23,34 +23,172 @@ using namespace mlir::torch;
 using namespace mlir::torch::Torch;
 
 namespace {
-#define DECL_CONVERT_ATENOP(aten_op)                                              \
-class ConvertAten##aten_op##Op : public OpConversionPattern<Aten##aten_op##Op> {  \
-public:                                                                           \
-  using OpConversionPattern::OpConversionPattern;                                 \
-  LogicalResult                                                                   \
-  matchAndRewrite(Aten##aten_op##Op op, ArrayRef<Value> operands,                 \
-                  ConversionPatternRewriter &rewriter) const override;            \
-}; 
+// These legalizations are for unary ops with only for FP datatypes.
+// There is no supported quantized integer mode for these.
+#define DEF_FULLCONV_FPONLY_UNARY_ATENOP(aten_op, tosa_op)                     \
```

Comment:
Hi Suraj, as a matter of programming style, the LLVM style guide (which we follow) generally prefers not to use macros, unless absolutely necessary. This really seems like a job for C++ templates to me. Can you rewrite this with templates? It might look something like this:

```
template <typename TorchOp, typename TosaOp>
class ConvertSimpleElementwiseOp : public OpConversionPattern<TorchOp> {
public:
  using OpConversionPattern::OpConversionPattern;
  LogicalResult
  matchAndRewrite(TorchOp op, ArrayRef<Value> operands,
                  ConversionPatternRewriter &rewriter) const override {
    TorchOp::Adaptor adaptor(operands);
    Value self = adaptor.self();
    auto selfTy = self.getType().cast<TensorType>();
    if (selfTy) {
      if (selfTy.getElementType().isa<mlir::FloatType>()) {
        rewriter.replaceOpWithNewOp<TosaOp>(
            op, getTypeConverter()->convertType(op.getType()), self);
        return success();
      } else {
        return op.emitError("Only FP type legalization supported");
      }
    } else {
      return op.emitError("Only Tensor types supported in TOSA");
    }
  }
};
```

Reply:
Yup, you read my mind - the ongoing refactoring intended to move this to templates. Perhaps easier in a followon push ?

Reply:
The original thinking was to keep aligned to TF->TOSA look and feel but 20 line macros are clearly an overkill.

Reply:
If you already have a patch stack built up around this, I'm ok doing the cleanup as a follow-on.

Reply:
Yes please, next round will template-ize this and also add @pashu123 's feedback , plus potentially a few more ops, since this structure makes the process of adding those common case ones quite trivial.

Reply:
:+1: